_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+### Setting the height of the content
+
+Now that we've calculated the height of the hidden element, we can use that to directly apply an inline style. This part is actually pretty easy.
+
+{% raw %}
+``` jsx
+<div
+  className="Drawer-content"
+  style={{ height: this.state.contentHeight }}
+>
+  {children}
+</div>
+```
+{% endraw %}
+
+We use an object that contains camelCased CSS attributes as React requires this for dealing with inline styles. For example, if we were working with `max-height` we would instead write the attribute above like: {% raw %}`style={{ maxHeight: this.state.contentHeight }}`{% endraw %}.
+
+With this our component sets the hidden content's height accurately.
+
+![Set Height Component](/img/2019/set-height.gif)
+
+## The dangers of nextSibling
+
+There's still one piece of the above that's bothering me. We're currently finding the element to calculate height based on the following chain of method calls `event.target.nextSibling.scrollHeight`. Well what happens if `nextSibling` doesn't exist? Or if we modify our component to include a new element between the button and the content?
+
+If `nextSibling` doesn't exist then calling `scrollHeight` on undefined is going to raise an exception. On the flipside, modifying the component means the content element will no longer be targeted as the next sibling.
+
+``` jsx
+return (
+  <div className={`Drawer ${drawerStyles}`}>
+    <button
+      className="Drawer-title"
+      onClick={this.handleTitleClick}
+    >
+      {title}
+    </button>
+
+    <div>In-between content section</div>
+
+    <div className="Drawer-content" {% raw %}style={{ height: this.state.contentHeight }}{% endraw %}>
+      {children}
+    </div>
+  </div>
+);
+```
+
+With the new `<div>In-between content section</div>` above we've broken our current height calculation. Now it looks like this:
+
+![Broken Drawer](/img/2019/broken-drawer.gif)
+
+It turns out that our component isn't very flexible to future changes.
+
+We can fix this by using React Refs.
+
+## Using Refs for accuracy
+
+<blockquote class="Info Info-right">
+  Refs provide a way to access DOM nodes or React elements created in the render method.
+  <cite>- <a href="https://reactjs.org/docs/refs-and-the-dom.html">reactjs.org</a></cite>
+</blockquote>
+
+Refs allow us to accurately target specific DOM element within the `render()` function. In our
+case, being able to target the content div via a ref is going provide greater flexibility to future changes
+while being less brittle. Here's how you would implement it:
+
```

Comment:
is going provide greater flexibility to future changes and will be less brittle.